### PR TITLE
Add ignore_if_prev_successes to copy plugin

### PR DESCRIPTION
**Pull request by TiagoGoddard**
Signed-off-by: Tiago Goddard <tiago.goddard@hotmart.com>

**Which issue(s) this PR fixes**: 
None

**What this PR does / why we need it**: 
This PR makes a change in the Copy Output plugin so you can inform the fluentd to ignore said copy if the previous output succeeded.

This enables having multiple nested outputs as backup of each other by doing:

```
    <store ignore_error>
      @type test
      name c0
    </store>
    <store ignore_if_prev_success ignore_error>
      @type test
      name c1
    </store>
    <store ignore_if_prev_success>
      @type test
      name c2
    </store>
```

**Docs Changes**:

Adding the defined ignore_if_prev_successes attribute to the tag makes the copy plugin ignore said store if the previous store succeed in pushing the output.

### CODE REVIEW

**Comment by repeatedly** [comment]
> the previous output succeeded.

This `succeeded` means data is stored into output's buffer, not write data to destination. Is this intended, right?

**Comment by TiagoGoddard** [comment]
> > the previous output succeeded.
> 
> This `succeeded` means data is stored into output's buffer, not write data to destination. Is this intended, right?

Yes, it isn't the responsibility of the copy plugin to assure the output does indeed write data to the destination, and it will raise the exception if it fails to do so if I'm not mistaken.

File: lib/fluent/plugin/out_copy.rb [hunk 1]
```diff
@@ -55,10 +57,15 @@ def process(tag, es)
...
-
+      success = []
       outputs.each_with_index do |output, i|
         begin
-          output.emit_events(tag, @copy_proc ? @copy_proc.call(es) : es)
+          if i > 0 && success[i - 1] && @ignore_if_prev_successes[i]
+            log.info "ignore copy because prev_success in #{output.plugin_id}", index: i
+          else
+            output.emit_events(tag, @copy_proc ? @copy_proc.call(es) : es)
+            success[i] = true;
```
[the remark this answers is not in the document]
Comment (TiagoGoddard) [reply lib/fluent/plugin/out_copy.rb]: Aggred, I'm too used to other languages.

File: lib/fluent/plugin/out_copy.rb [hunk 2]
```diff
@@ -55,10 +57,15 @@ def process(tag, es)
         }
         es = m
       end
-
+      success = []
```
[the remark this answers is not in the document]
Comment (TiagoGoddard) [reply lib/fluent/plugin/out_copy.rb]: Aggred, I'm too used to other languages.

File: lib/fluent/plugin/out_copy.rb [hunk 3]
```diff
@@ -55,10 +57,15 @@ def process(tag, es)
...
       outputs.each_with_index do |output, i|
         begin
-          output.emit_events(tag, @copy_proc ? @copy_proc.call(es) : es)
+          if i > 0 && success[i - 1] && @ignore_if_prev_successes[i]
+            log.info "ignore copy because prev_success in #{output.plugin_id}", index: i
```
[the remark this answers is not in the document]
Comment (TiagoGoddard) [reply lib/fluent/plugin/out_copy.rb]: Aggred, it really will push a lot of info logs

File: lib/fluent/plugin/out_copy.rb [hunk 4]
```diff
@@ -40,6 +41,7 @@ def configure(conf)
       @copy_proc = gen_copy_proc
       @stores.each { |store|
         @ignore_errors << (store.arg == 'ignore_error')
+        @ignore_if_prev_successes << (store.arg == 'ignore_if_prev_success')
```
[the remark this answers is not in the document]
Comment (TiagoGoddard) [reply lib/fluent/plugin/out_copy.rb]: Aggred

File: test/plugin/test_out_copy.rb [hunk 5]
```diff
@@ -217,5 +217,37 @@ def test_ignore_error
...
+      @type test
+      name c1
+    </store>
+    <store ignore_if_prev_success>
+      @type test
+      name c2
+    </store>
+  ]
+
+  def test_ignore_if_prev_success
```
Comment (repeatedly) [inline test/plugin/test_out_copy.rb:236]: Does this code test `ignore_if_prev_success` feature?
This code is copied from `ignore_error` and no modification, so this code tests `ignore_error`.
This code should test 2nd/3rd `<store>`'s `process` is not called.

**Comment by cosmo0920** [comment]
Closing this in favor of https://github.com/fluent/fluentd/pull/3287.

**Pull request closed without merging**